src/closest.c

```diff
@@ -70,91 +20,41 @@ int leftmost(double x, double* ptable, int low, int n) {
-            cur = 0;
-        else if (px[i] >= ptable[ntable - 1])
-            cur = ntable - 1;
-        else {
-            low = leftmost(px[i], ptable, low, ntable);
-            cur = (px[i] - ptable[low] <= ptable[low + 1] - px[i]) ?
-                low : low + 1;
-        }
-
-        /* cur + 1 is needed here to translate between R's and C's indices */
-        pout[i] = fabs(px[i] - ptable[cur]) <= ptolerance[i] ?
-            cur + 1 : asInteger(nomatch);
-  }
-
-  UNPROTECT(1);
-  return(out);
-}
+    const unsigned int inomatch = asInteger(nomatch);
+    unsigned int j = 1;
 
-/**
- * Find closest value to table, keep just closest duplicate.
- * This function should be reused in the outer join C implementation.
- *
- * \param pout pointer to output array.
- * \param px pointer to arrays with key values to look for.
- * \param nx number of elements in `px`.
- * \param ptable table/haystack where to look for.
- * \param ntable number of elements in `ptable`.
- * \param ptolerance pointer to tolerance array that stores the tolerance
- * to be accepted as match, has to be of length == length(x).
- * \param nomatch value that should be returned if a key couldn't be matched.
- * \return index the closest element
- *
- * \note x and table have to be sorted increasingly and not containing any NA.
- */
-void closest_dup_closest(int *pout, double *px, int nx,
-                         double *ptable, int ntable,
-                         double *ptolerance, int nomatch) {
-
-    int low = 0, cur = 0;
-    int last = R_NegInf;
-    double absdiff, lastdiff = R_PosInf;
-
-    for (int i = 0; i < nx; ++i) {
-        if (px[i] < ptable[0])
-            cur = 0;
-        else if (px[i] >= ptable[ntable - 1])
-            cur = ntable - 1;
-        else {
-            low = leftmost(px[i], ptable, low, ntable);
-            cur = (px[i] - ptable[low] <= ptable[low + 1] - px[i]) ?
-                low : low + 1;
-        }
-
-        absdiff = fabs(px[i] - ptable[cur]);
-        if (absdiff <= ptolerance[i]) {
-            if (last == cur) {
-                if (absdiff < lastdiff && last == cur) {
-                    pout[i] = cur + 1;
-                    pout[i - 1] = nomatch;
-                    lastdiff = absdiff;
-                } else
-                    pout[i] = nomatch;
-            } else {
-                pout[i] = cur + 1;
-                last = cur;
-                lastdiff = absdiff;
-            }
+    double prevdiff = R_PosInf, nextdiff = R_PosInf;
+
+    for (unsigned int i = 0; i < nx; ++i) {
+        while(j < ntable1 && ptable[j] < px[i])
+            ++j;
+
+        /* fabs should be just needed for the first element */
+        prevdiff = fabs(px[i] - ptable[j - 1]);
+        nextdiff = fabs(ptable[j] - px[i]);
+
+        if (prevdiff <= ptolerance[i] || nextdiff <= ptolerance[i]) {
+            if (prevdiff <= nextdiff)
+                pout[i] = j;
+            else
+                pout[i] = ++j;
```

Comment:
just a curiosity - is `j++` different than `++j`?

Reply:
`j++` increments `j` but returns `j` (reading as returning `j` and incrementing it afterwards) vs `++j` increments `j` and returns `j + 1` (reading as incrementing first and returning the result). So `pout[i] = ++j` is `j = j + 1; pout[i] = j`. `pout[i] = j++` would be `pout[i] = j; j = j + 1;`.